File: src/dp/_416_can_partition.rs
struct Solution;

impl Solution {
    pub fn can_partition(nums: Vec<i32>) -> bool {
        if nums.is_empty() {
            return false;
        }
        let n = nums.len();
        let sum: i32 = nums.iter().sum();
        if sum % 2 != 0 {
            return false;
        }
        let target = sum / 2;
        let mut dp = vec![0; (target + 1) as usize];
        for i in 0..n {
            for j in (nums[i]..=target).rev() {
                dp[j as usize] = dp[j as usize].max(dp[(j - nums[i]) as usize] + nums[i]);
            }
        }
        dp[target as usize] == target
    }
}

#[test]
fn test(){
    let a = Solution::can_partition(vec![1,2,5]);
    assert_eq!(a, false);
}

File: hot-100/416_canPartition.go
package hot_100

func canPartition(nums []int) bool {
	sum := 0
	for _, n := range nums {
		sum += n
	}
	if sum%2 != 0 {
		return false
	}
	target := sum / 2
	dp := make([]int, target+1)
	for _, num := range nums {
		for i := target; i >= num; i-- {
			if dp[i] < dp[i-num]+num {
				dp[i] = dp[i-num] + num
			}
		}
	}
	return dp[target] == target
}
